pkg/migration/controllers/migration.go

```diff
@@ -1697,109 +1700,149 @@ func (m *MigrationController) applyResources(
+								if err != nil && errors.IsNotFound(err) {
+									break
+								}
+								createTime := obj.GetCreationTimestamp()
+								if deleteStart.Before(&createTime) {
+									log.MigrationLog(migration).Warnf("Object[%v] got re-created after deletion. So, Ignore wait. deleteStart time:[%v], create time:[%v]",
+										obj.GetName(), deleteStart, createTime)
+									break
+								}
+								log.MigrationLog(migration).Warnf("Object %v still present, retrying in %v", metadata.GetName(), deletedRetryInterval)
+								time.Sleep(deletedRetryInterval)
 							}
-							log.MigrationLog(migration).Warnf("Object %v still present, retrying in %v", metadata.GetName(), deletedRetryInterval)
-							time.Sleep(deletedRetryInterval)
+							_, err = dynamicClient.Create(context.TODO(), unstructured, metav1.CreateOptions{})
 						}
-						_, err = dynamicClient.Create(context.TODO(), unstructured, metav1.CreateOptions{})
 					}
 				}
+				// Retry a few times for Unauthorized errors
+				if err != nil && errors.IsUnauthorized(err) && retries < maxApplyRetries {
+					retries++
+					continue
+				}
+				break
 			}
-			// Retry a few times for Unauthorized errors
-			if err != nil && errors.IsUnauthorized(err) && retries < maxApplyRetries {
-				retries++
-				continue
+			if err != nil {
+				m.updateResourceStatus(
+					migration,
+					o,
+					stork_api.MigrationStatusFailed,
+					fmt.Sprintf("Error applying resource: %v", err))
+			} else {
+				m.updateResourceStatus(
+					migration,
+					o,
+					stork_api.MigrationStatusSuccessful,
+					"Resource migrated successfully")
 			}
-			break
+			errorChan <- nil
 		}
-		if err != nil {
-			m.updateResourceStatus(
-				migration,
-				o,
-				stork_api.MigrationStatusFailed,
-				fmt.Sprintf("Error applying resource: %v", err))
-		} else {
-			m.updateResourceStatus(
-				migration,
-				o,
-				stork_api.MigrationStatusSuccessful,
-				"Resource migrated successfully")
+	}
+
+	return m.parallelWorker(worker, updatedObjects, true)
+}
+
+func (m *MigrationController) parallelWorker(
+	worker func(<-chan runtime.Unstructured, chan<- error),
+	objects []runtime.Unstructured,
+	shuffle bool,
+) error {
+	numObjects := len(objects)
+	objectChan := make(chan runtime.Unstructured, 100)
+	errorChan := make(chan error, 100)
+
+	if shuffle {
+		// Shuffle Object order before applying so we can get parallelism between resource types
+		rand.Seed(time.Now().UnixNano())
+		rand.Shuffle(len(objects), func(i, j int) { objects[i], objects[j] = objects[j], objects[i] })
+	}
+
+	logrus.Infof("Updating %v objects with %v parallel workers", numObjects, m.migrationMaxThreads)
+	for w := 0; w < m.migrationMaxThreads; w++ {
+		go worker(objectChan, errorChan)
+	}
+
+	go func() {
+		for _, o := range objects {
+			objectChan <- o
+		}
+		close(objectChan)
+	}()
+
+	for result := 0; result < numObjects; result++ {
+		workerErr := <-errorChan
+		if workerErr != nil {
+			return workerErr
```

Comment:
We need to let the other goroutines know that we are not going to be processing any more updates when one of them returns an error. In this way you can deterministically cleanup those goroutines.

Reply:
Changed to use unbuffered channels, so when we find an error or finish results we close objectChan and the workers will exit immediately after their current piece of work. This should address both comments.